src/qutip_qip/circuit/text_renderer.py

```diff
@@ -0,0 +1,497 @@
+        node_conn = "─" * (width // 2) + "▇" + "─" * (width // 2 - 1)
+        self._adjust_layer(wire_list_control, xskip)
+
+        for wire in wire_list_control:
+            if wire not in gate.targets:
+                if wire in gate.controls:
+                    # check if the control wire is the first or last control wire.
+                    # used in cases of multiple control wires
+                    if (
+                        wire == wire_list_control[0]
+                        or wire == wire_list_control[-1]
+                    ):
+                        self._render_strs["mid_frame"][wire] += node_conn
+                        self._render_strs["top_frame"][wire] += (
+                            bar_conn if is_top_closer else " " * len(bar_conn)
+                        )
+                        self._render_strs["bot_frame"][wire] += (
+                            bar_conn
+                            if not is_top_closer
+                            else " " * len(bar_conn)
+                        )
+                    else:
+                        self._render_strs["top_frame"][wire] += bar_conn
+                        self._render_strs["mid_frame"][wire] += node_conn
+                        self._render_strs["bot_frame"][wire] += bar_conn
+                else:
+                    self._render_strs["top_frame"][wire] += bar_conn
+                    self._render_strs["mid_frame"][wire] += mid_bar_conn
+                    self._render_strs["bot_frame"][wire] += bar_conn
+
+    def _update_swap_gate(self, gate: Gate):
+        """
+        Update the render strings for the SWAP gate
+        """
+
+        wire_list = list(range(min(gate.targets), max(gate.targets) + 1))
+        layer = max(len(self._layer_list[i]) for i in wire_list)
+        xskip = self._get_xskip(wire_list, layer)
+        self._adjust_layer(wire_list, xskip)
+
+        width = 4 * ceil(self.style.gate_pad) + 1
+        cross_conn = "─" * (width // 2) + "╳" + "─" * (width // 2)
+        bar_conn = " " * (width // 2) + "│" + " " * (width // 2)
+        mid_bar_conn = "─" * (width // 2) + "│" + "─" * (width // 2)
+        self._manage_layers(width, wire_list, layer, xskip)
+
+        for wire in wire_list:
+            if wire == wire_list[-1]:
+                self._render_strs["top_frame"][wire] += " " * len(bar_conn)
+                self._render_strs["mid_frame"][wire] += cross_conn
+                self._render_strs["bot_frame"][wire] += bar_conn
+            elif wire == wire_list[0]:
+                self._render_strs["top_frame"][wire] += bar_conn
+                self._render_strs["mid_frame"][wire] += cross_conn
+                self._render_strs["bot_frame"][wire] += " " * len(bar_conn)
+            else:
+                self._render_strs["top_frame"][wire] += bar_conn
+                self._render_strs["mid_frame"][wire] += mid_bar_conn
+                self._render_strs["bot_frame"][wire] += bar_conn
+
+    def layout(self):
```

Comment:
What is the difference between `layout` and `print_circuit`?

Reply:
The `layout` function iterates through the gates and updates the `_render_strs` list of strings, while `print_circuit` handles displaying the `_render_strs` in the correct order, either in the terminal or within an IPython environment.